### DISCUSSION TITLE
[Image] Use `sizes` property for generating `srcset` when available

### DISCUSSION THREAD

**Discussion by kripod** [vercel/next.js#18920] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

The `next/image` component is great for rendering images at optimal sizes. However, with a `sizes` attribute provided, we could infer all the possible dimensions rather than estimating them by using `deviceSizes`.

## Describe the solution you'd like

Additional `imageSizes` should be inferred for each image from their `sizes` prop (when given). For example:

```jsx
<Image
  src="/assets/avatar.png"
  sizes="(min-width: 1024px) 120px, (min-width: 640px) 96px, 72px"
  width={120}
  height={120}
  className="rounded-full"
/>
```

The image above would only be displayed with the following resolutions:

- 72x72px on mobile
- 96x96px when viewport width ≥640px
- 120x120px when viewport width ≥1024px

Based on screen pixel density, the following sizes would be useful:

- 72x72px, 144x144px and 216x216px on mobile
- 96x96px and 192x192px when viewport width ≥640px
- 120x120px and 240x240px when viewport width ≥1024px

(I omitted `3x` dimensions on purpose for larger displays.)

## Describe alternatives you've considered

Using `imageSizes` on its own, but it could have unintended side effects on media uploaded by users.

**Comment by amannn** [comment]
I think I have a similar situation: I have an image that has a fluid width on mobile and therefore needs `layout="responsive"`. On desktop the image has a fixed with of `400`. Now I have another image with `layout="responsive"` that has a maximum width of `1000` on desktop.

Based on this, I think I need this configuration:

```
deviceSizes: [..., 400, 1000]
```

This would however cause the first image which has a maximum width of 400 to be requested at a width of 1000 on desktop, which is far too large.

What I'd need is a case-by-case configuration for responsive images.

I thought about reading the current breakpoint in JS and then changing the `layout` of the first image on desktop. The issue with this is however, that the JS has to initialize first, which will cause an unused image to be loaded when the SSR'd markup is received by the browser.

To give some context to the use case: A single column layout on mobile which grows in width is changed into a two column layout on desktop, where the sidebar has a fixed width of 400.

**Comment by OskarGroth** [comment]
2 years later, this is still a huge problem: https://github.com/vercel/next.js/issues/37265
